contracts/src/Chainvoice.sol

```diff
@@ -1,104 +1,190 @@
 
-        // Transfer Ether to the sender
-        (bool success,) = payable(invoice.from).call{value: msg.value}("");
+        uint256 feeAmountInNativeCurrency = usdToNativeCurrencyConversion();
+        require(
+            msg.value >= invoice.amountDue + feeAmountInNativeCurrency,
+            "Payment must cover the invoice amount and fee"
+        );
+        uint256 amountToRecipient = msg.value - feeAmountInNativeCurrency;
+        (bool success, ) = payable(invoice.from).call{value: amountToRecipient}(
+            ""
+        );
         require(success, "Payment transfer failed");
+        accumulatedFees += feeAmountInNativeCurrency;
         invoice.isPaid = true;
-        emit InvoicePaid(invoiceId, invoice.from, msg.sender, msg.value);
     }
-    /**
-     * @dev Get all invoices sent by the caller
-     * @return An array of InvoiceDetails for invoices sent by the caller
-     */
-    function getMySentInvoices() external view returns (InvoiceDetails[] memory) {
+
+    function getMySentInvoices()
+        external
+        view
+        returns (InvoiceDetails[] memory, ItemData[][] memory)
+    {
         return _getInvoices(sentInvoices[msg.sender]);
     }
 
-    /**
-     * @dev Get all invoices received by the caller
-     * @return An array of InvoiceDetails for invoices received by the caller
-     */
-    function getMyReceivedInvoices(address add) external view returns (InvoiceDetails[] memory) {
+    function getMyReceivedInvoices(
+        address add
+    ) external view returns (InvoiceDetails[] memory, ItemData[][] memory) {
         return _getInvoices(receivedInvoices[add]);
     }
 
-    /**
-     * @dev Internal function to fetch invoices by IDs
-     * @param invoiceIds Array of invoice IDs to fetch
-     * @return An array of InvoiceDetails
-     */
-    function _getInvoices(uint256[] storage invoiceIds) internal view returns (InvoiceDetails[] memory) {
-        InvoiceDetails[] memory userInvoices = new InvoiceDetails[](invoiceIds.length);
+    function _getInvoices(
+        uint256[] storage invoiceIds
+    ) internal view returns (InvoiceDetails[] memory, ItemData[][] memory) {
+        InvoiceDetails[] memory userInvoices = new InvoiceDetails[](
+            invoiceIds.length
+        );
+        ItemData[][] memory items = new ItemData[][](invoiceIds.length);
 
         for (uint256 i = 0; i < invoiceIds.length; i++) {
             userInvoices[i] = invoices[invoiceIds[i]];
+            items[i] = itemDatas[invoiceIds[i]];
         }
 
-        return userInvoices;
+        return (userInvoices, items);
+    }
+
+    function setTreasuryAddress(address add) public OnlyOwner {
+        require(add != address(0), "Treasury Address cannot be equal to zero");
+        treasuryAddress = add;
+    }
+
+    function getTreasuryAddress() public view returns (address) {
+        return treasuryAddress;
+    }
+
+    function setFeeAmount(uint16 fee) public OnlyOwner {
+        feeAmountInUSD = fee * 1e18;
```

Comment:
Change the name of this variable to `fee` and do not multiply by `1e18`.

The fee should be in weis.

Multiplying by `1e18` prevents us from having decimal fees. For example, on Ethereum, the minimum non-zero fee would be 1e18 weis, which is equal to 1 ETH, which is equal to 2000+ USD!